### PR TITLE
Do not emit shadow copied for inout parameters

### PR DESCRIPTION
**Pull request by shajrawi**
radar rdar://problem/28434323

SILGen has no reason to insert shadow copies for inout parameters any more. They cannot be captured. We still emit these copies. Sometimes deshadowing removes them, but sometimes it does not.

In this PR we just avoid emitting the copies and remove the deshadowing pass.

This PR chery-picked some of @dduan work and built on top of it.

### CODE REVIEW

**Comment by shajrawi** [comment]
@atrick Can you please review?

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

File: lib/SILGen/SILGenProlog.cpp [hunk 1]
```diff
@@ -262,19 +262,10 @@ struct ArgumentInitHelper {
         gen.B.createDebugValueAddr(loc, address, {vd->isLet(), ArgNo});
         return;
       }
+      gen.VarLocs[vd] = SILGenFunction::VarLoc::get(argrv.getValue());
+      assert(argrv.getType().isAddress() && "expected inout to be address");
+      gen.B.createDebugValueAddr(loc, argrv.getValue(), {vd->isLet(), ArgNo});
```
Comment (atrick) [inline lib/SILGen/SILGenProlog.cpp]: The BuiltinUnsafeValueBufferType case above passes the unmanaged address to DebugValueAddr. This new default case passes a the argv value directly to DebugValueAddr. Why is there any difference?

**Review by atrick — CHANGES_REQUESTED** [review]
Can you identify some unit tests that handle captured inout, and are you convinced they do the right thing? Can you add a test from the bug report to verify that local functions capturing inout are now handled the same as anonymous closures?

Otherwise this looks fine to me.

**Comment by lattner** [comment]
Awesome, thank you for working on this Joe, I look forward to this all getting cleaner and simpler.

File: lib/SILGen/SILGenProlog.cpp (hunk 1 above)
Comment (shajrawi) [inline lib/SILGen/SILGenProlog.cpp]: Combined debug value creation + added unit tests as requested.

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
Linux smoke test failing with the usual non-PR related error - rerunning:
/home/buildnode/disk2/workspace/swift-PR-Linux-smoke-test/branch-master/swiftpm/Tests/XcodeprojTests/XcodeProjectModelTests.swift:12:8: error: module file was created by an older version of the compiler; rebuild 'TestSupport' and try again: /home/buildnode/disk2/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/swiftpm-linux-x86_64/debug/TestSupport.swiftmodule

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by CodaFi** [comment]
Before this is merged, could you please squash @dduan's commits?

**Comment by dduan** [comment]
Yes please squash...these were never meant to be public :)

**Comment by shajrawi** [comment]
I intended to auto-squash everything into one commit on merge - all commits are about the same thing :)

**Review by atrick — APPROVED** [review]

**Comment by swiftix** [comment]
@shajrawi Have you forgotten to squash the commits?

**Comment by shajrawi** [comment]
@swiftix I squashed it - it is a single commit on mainline, no? (I clicked squash and merge In the PR)

**Comment by swiftix** [comment]
@shajrawi Ah, OK. I assumed you you'd squash and push --force on the PR first and then merge. This is the usual way to it, because then the PR looks the same as the final merged set of commits.

**Pull request merged**